rust/nasl-interpreter/src/built_in_functions/string.rs

```diff
@@ -0,0 +1,319 @@
+///
+/// If this function retrieves anything but a string it returns NULL
+fn toupper(_: &str, _: &dyn Sink, register: &Register) -> Result<NaslValue, FunctionError> {
+    let positional = resolve_positional_arguments(register);
+    Ok(match positional.get(0) {
+        Some(NaslValue::String(x)) => x.to_uppercase().into(),
+        _ => NaslValue::Null,
+    })
+}
+
+/// NASL function to return lowercase equivalent of a given string
+///
+/// If this function retrieves anything but a string it returns NULL
+fn tolower(_: &str, _: &dyn Sink, register: &Register) -> Result<NaslValue, FunctionError> {
+    let positional = resolve_positional_arguments(register);
+    Ok(match positional.get(0) {
+        Some(NaslValue::String(x)) => x.to_lowercase().into(),
+        _ => NaslValue::Null,
+    })
+}
+
+/// NASL function to return the length of string
+///
+/// If this function retrieves anything but a string it returns 0
+fn strlen(_: &str, _: &dyn Sink, register: &Register) -> Result<NaslValue, FunctionError> {
+    let positional = resolve_positional_arguments(register);
+    Ok(match positional.get(0) {
+        Some(NaslValue::String(x)) => x.len().into(),
+        _ => 0_i64.into(),
+    })
+}
+
+/// NASL function to return a substr of a string.
+///
+/// The first positional argument is the *string* to get the slice from.
+/// As a second positional argument an *int* that contains the start index for the slice is required.
+/// The optional third positional argument is an *int* and contains the end index for the slice.
+/// If not given it is set to the end of the string.
+/// If the start integer is higher than the value of the string NULL is returned.
+fn substr(_: &str, _: &dyn Sink, register: &Register) -> Result<NaslValue, FunctionError> {
+    let positional = resolve_positional_arguments(register);
+    if positional.len() < 2 {
+        return Ok(NaslValue::Null);
+    }
+    // we checked previously if the size is sufficient
+    unsafe {
+        let s = positional.get_unchecked(0).to_string();
+        let lidx: i64 = positional.get_unchecked(1).into();
+        if lidx as usize > s.len() {
+            return Ok(NaslValue::Null);
+        }
+        Ok(match positional.get(2) {
+            Some(nv) => {
+                let ridx: i64 = nv.into();
+                (&s[lidx as usize..ridx as usize]).into()
+            }
+            _ => (&s[lidx as usize..]).into(),
+        })
+    }
+}
+
+/// NASL function to return  a hex presentation of given string as a positional argument.
+///
+/// If the positional arguments are empty it reutnrns NaslValue::Null.
+/// It only uses the first positional argument and when it is not a NaslValue:String than it returns NaslValue::Null.
+fn hexstr(_: &str, _: &dyn Sink, register: &Register) -> Result<NaslValue, FunctionError> {
+    let positional = resolve_positional_arguments(register);
+    // we already checked that positional has at least one argument
```

Comment:
I am pretty sure this is not true, as resolve_positional_argument should return an empty vec, if no positional arguments were given. In the end we will return `NaslValue::NULL` in this case, as it is covered within the followed match block.

Reply:
yes, that comment is a leftover of a previous solution. Should be deleted